# LABORATORIO 2
### Marianna Flores 20180040

In [1]:
import numpy as np
import pandas as pd
heroes = pd.read_csv("heroes_information.csv")

In [2]:
###instalar desde anaconda navigator######
from pandasql import *

In [3]:
pysqldf = lambda q: sqldf(q,globals())


### Contenido tabla 

In [4]:
query2 = " SELECT * FROM heroes limit 10"

In [5]:
pysqldf(query2)

id           name  Gender Eye color               Race Hair color  Height  \
0   0         A-Bomb    Male    yellow              Human    No Hair   203.0   
1   1     Abe Sapien    Male      blue      Icthyo Sapien    No Hair   191.0   
2   2       Abin Sur    Male      blue            Ungaran    No Hair   185.0   
3   3    Abomination    Male     green  Human / Radiation    No Hair   203.0   
4   4        Abraxas    Male      blue      Cosmic Entity      Black   -99.0   
5   5  Absorbing Man    Male      blue              Human    No Hair   193.0   
6   6    Adam Monroe    Male      blue               None      Blond   -99.0   
7   7   Adam Strange    Male      blue              Human      Blond   185.0   
8   8       Agent 13  Female      blue               None      Blond   173.0   
9   9      Agent Bob    Male     brown              Human      Brown   178.0   

           Publisher Skin color Alignment  Weight  
0      Marvel Comics       None      good   441.0  
1  Dark Horse Comics       blue      good    65.0  
2          DC Comics        red      good    90.0  
3      Marvel Comics       None       bad   441.0  
4      Marvel Comics       None       bad   -99.0  
5      Marvel Comics       None       bad   122.0  
6       NBC - Heroes       None      good   -99.0  
7          DC Comics       None      good    88.0  
8      Marvel Comics       None      good    61.0  
9      Marvel Comics       None      good    81.0


#### 1. Nombre primer superheroe

In [6]:
q1 = " SELECT * FROM heroes limit 1"

In [7]:
pysqldf(q1)

id    name Gender Eye color   Race Hair color  Height      Publisher  \
0   0  A-Bomb   Male    yellow  Human    No Hair   203.0  Marvel Comics   

  Skin color Alignment  Weight  
0       None      good   441.0

El primer superheroe es A-Bomb

#### 2. Cuantas casas publicadoras existen

In [8]:
q2 = "SELECT COUNT(DISTINCT Publisher) as cantidad FROM heroes"

In [9]:
pysqldf(q2)

cantidad
0        24

Hay 24 casas publicadoras

#### 3. Cuantos superheroes miden mas de 2m

In [10]:
q3 = " SELECT COUNT(*) as cantidad FROM heroes WHERE Height > 200"

In [11]:
pysqldf(q3)

cantidad
0        59

59 heroes miden mas de 2m

#### 4. Cuantos son humanos y miden mas de 2m

In [12]:
q4 = " SELECT COUNT(*) as cantidad FROM heroes WHERE Height > 200 AND Race = 'Human'"

In [13]:
pysqldf(q4)

cantidad
0        12

Hay 12 superheroes que son humanos y miden mas de 2m

#### 5. Cuantos pesan mas de 100lbs y menos de 200lbs

In [14]:
q5 = " SELECT COUNT(*) as cantidad FROM heroes WHERE Weight BETWEEN 100 AND 200"

In [15]:
pysqldf(q5)

cantidad
0        98

Hay 98 heroes con pesos entre 100 y 200

#### 6. Cuantos tienen los ojor color azul o rojo

In [16]:
q6 = " SELECT COUNT(*) FROM heroes WHERE [Eye color]='blue' OR [Eye color]='red'"

In [17]:
pysqldf(q6)

COUNT(*)
0       271

Hay 271 superheroes con ojos azules o rojos

#### 7. Cuántos superhéroes son Human, Mutant y tienen el pelo color Verde o son Vampiros con pelo Negro

In [18]:
q7 = " SELECT COUNT(*) FROM heroes WHERE (([Race]='Human' OR [Race]='Mutant') AND [Hair color]='Green') OR ([Race]='Vampire' AND [Hair color]='Black')"

In [19]:
pysqldf(q7)

COUNT(*)
0         5

Hay 5 superheroes que son humano o mutante con pelo verde, o vampiro con pelo negro

#### 8.Cuál es el nombre del primer superhéroe si ordeno la tabla por raza en orden descendente?

In [20]:
q8 = "SELECT * FROM heroes ORDER BY Race Desc limit 1"

In [21]:
pysqldf(q8)

id            name Gender Eye color    Race Hair color  Height  Publisher  \
0  611  Solomon Grundy   Male     black  Zombie      White   279.0  DC Comics   

  Skin color Alignment  Weight  
0       None       bad   437.0

El primer nombre es Solomon Grundy

#### 9. Cuántos superhéroes son de género masculino y cuántas de género femenino?

In [22]:
qf = "SELECT COUNT(*) as 'cantidad mujeres' FROM heroes WHERE Gender = 'Female'"
qm = "SELECT COUNT(*) as 'cantidad hombres' FROM heroes WHERE Gender = 'Male'"

In [23]:
pysqldf(qf)

cantidad mujeres
0               200

In [24]:
pysqldf(qm)

cantidad hombres
0               505

Hay 200 mujeres y 505 hombres

#### 10. Cuántas casas publicadoras tienen más de 15 superhéroes?

In [25]:
q10 = "SELECT COUNT(*) as cantidad FROM heroes GROUP BY Publisher HAVING COUNT(*)>15"

In [26]:
pysqldf(q10)

cantidad
0       215
1        18
2       388
3        19

Hay 4 casas publicadoras con mas de 15 superheroes